core/src/main/java/nl/nn/adapterframework/pipes/JwtPipe.java

```diff
@@ -0,0 +1,168 @@
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package nl.nn.adapterframework.pipes;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Date;
+import java.util.Map;
+import java.util.Objects;
+
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JOSEObjectType;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSSigner;
+import com.nimbusds.jose.KeyLengthException;
+import com.nimbusds.jose.crypto.MACSigner;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.JWTClaimsSet.Builder;
+import com.nimbusds.jwt.SignedJWT;
+
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import nl.nn.adapterframework.core.ParameterException;
+import nl.nn.adapterframework.core.PipeLineSession;
+import nl.nn.adapterframework.core.PipeRunException;
+import nl.nn.adapterframework.core.PipeRunResult;
+import nl.nn.adapterframework.doc.Category;
+import nl.nn.adapterframework.parameters.ParameterList;
+import nl.nn.adapterframework.parameters.ParameterValueList;
+import nl.nn.adapterframework.stream.Message;
+import nl.nn.adapterframework.util.CredentialFactory;
+
+/**
+ * Creates a JWT
+ *
+ * @author Niels Meijer
+ * @since 7.9
+ */
+@Category("Basic")
+public class JwtPipe extends FixedForwardPipe {
+	private static final String SHARED_SECRET_PARAMETER_NAME = "sharedSecret";
+
+	private JWSHeader jwtHeader;
+	private JWSSigner globalSigner;
+
+	private String sharedSecret;
+	private String sharedSecretAlias;
+	private int expirationTime = 600;
+
+	@Override
+	public void configure() throws ConfigurationException {
+		parameterNamesMustBeUnique = true;
+		super.configure();
+
+		jwtHeader = new JWSHeader.Builder(JWSAlgorithm.HS256).type(JOSEObjectType.JWT).build();
+
+		if(StringUtils.isNotEmpty(sharedSecret) || StringUtils.isNotEmpty(sharedSecretAlias)) {
+			try {
+				CredentialFactory credentialFactory = new CredentialFactory(sharedSecretAlias, null, () -> sharedSecret);
+				globalSigner = new MACSigner(credentialFactory.getPassword());
+			} catch (KeyLengthException e) {
+				throw new ConfigurationException("invalid shared key", e);
+			}
+		}
+
+		if(globalSigner == null && getParameterList().findParameter(SHARED_SECRET_PARAMETER_NAME) == null) {
+			throw new ConfigurationException("must either provide a [sharedSecret] (alias) or parameter");
+		}
+	}
+
+	@Override
+	public PipeRunResult doPipe(Message message, PipeLineSession session) throws PipeRunException {
+		final JWSSigner signer;
+		Builder claimsSetBuilder = new JWTClaimsSet.Builder();
+
+		Map<String, Object> parameterMap = getParameterValueMap(message, session);
+		if(parameterMap != null) {
+			Object sharedKey = parameterMap.remove(SHARED_SECRET_PARAMETER_NAME);
```

Comment:
If there is a map, but this param is empty, then it uses the attribute.

Reply:
Oh, I don't like that. The code here does already have an "else" that uses the attribute `globalSigner` and then the method `getSigner()` does it again, but less visible.

Since the map is now always non-null I think it is cleaner to check the map for presence of the parameter. If this parameter is not present, it will then always and clearly fall back to the global signer attribute.